### PR TITLE
Add DirectByteBufferPool for WAL

### PR DESCRIPTION
**Pull request by JackieTien97**
There are two problems in the process of current wal flushing.

- Because wal uses FileChannel while flushing, it involves one more memory copy. FileChannel will copy the contents of HeapByteBuffer to DirectByteBuffer before calling native write method, and then call native write with DirectByteBuffer as a parameter, which will cause An extra memory copy operation.
- Although 0.12 uses the thread pool to complete asynchronous flushing, to a certain extent, solves the problem of the surge of non-heap memory caused by the wal of 0.11, but when the non-heap memory is released is still depends on the jvm (depends on when gc is performed), jvm cannot know whether non-heap memory should be released in time, so it may also cause non-heap memory to not be released in time, but we are clear about that.

So, I want to use DirectByteBufferPool to solve the above two questions. The detailed design documents is on the confluence(https://cwiki.apache.org/confluence/display/IOTDB/DirectByteBuffer+Pool+for+WAL).

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupInfo.java [hunk 1]
```diff
@@ -94,13 +100,46 @@ public void setLastReportedSize(long size) {
...
+      return this::walSupplier;
+    }
+  }
+
+  @TestOnly
+  private ByteBuffer[] walSupplier() {
+    ByteBuffer[] buffers = new ByteBuffer[2];
+    buffers[0] = ByteBuffer
+        .allocateDirect(IoTDBDescriptor.getInstance().getConfig().getWalBufferSize() / 2);
+    buffers[0] = ByteBuffer
```
Comment (HTHou) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupInfo.java:126]: Should be buffers[1]?

**Review by jt2594838 — APPROVED** [review]

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 2]
```diff
@@ -261,6 +268,79 @@ public void setReady(boolean ready) {
...
+    synchronized (walByteBufferPool) {
+      int expectedSize =
+          (workSequenceTsFileProcessors.size() + workUnsequenceTsFileProcessors.size()) * 2;
+      while (expectedSize < currentWalPoolSize && !walByteBufferPool.isEmpty()) {
+        MmapUtil.clean((MappedByteBuffer) walByteBufferPool.removeLast());
+        MmapUtil.clean((MappedByteBuffer) walByteBufferPool.removeLast());
+        currentWalPoolSize -= 2;
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (JackieTien97) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java]: fix

File: server/src/main/java/org/apache/iotdb/db/utils/MmapUtil.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+package org.apache.iotdb.db.utils;
+
+import io.netty.util.internal.PlatformDependent;
+import java.nio.MappedByteBuffer;
+
+public class MmapUtil {
+
+  public static void clean(MappedByteBuffer mappedByteBuffer) {
+    if (mappedByteBuffer == null || !mappedByteBuffer.isDirect() || mappedByteBuffer.capacity()== 0)
+      return;
```
[the remark this answers is not in the document]
Comment (JackieTien97) [reply server/src/main/java/org/apache/iotdb/db/utils/MmapUtil.java]: fix

File: server/src/main/java/org/apache/iotdb/db/writelog/node/ExclusiveWriteLogNode.java [hunk 4]
```diff
@@ -92,6 +90,11 @@ public ExclusiveWriteLogNode(String identifier) {
     }
   }
 
+  public void initBuffer(ByteBuffer[] byteBuffers) {
+    this.logBufferWorking = byteBuffers[0];
+    this.logBufferIdle = byteBuffers[1];
+  }
```
[the remark this answers is not in the document]
Comment (JackieTien97) [reply server/src/main/java/org/apache/iotdb/db/writelog/node/ExclusiveWriteLogNode.java]: fix

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 5]
```diff
@@ -320,6 +331,10 @@ public void releaseWalBuffer(ByteBuffer[] byteBuffers) {
       byteBuffer.clear();
     }
     synchronized (walByteBufferPool) {
+      // if the pool is empty before, update the time
+      if (walByteBufferPool.isEmpty()) {
+        timeWhenPoolNotEmpty = System.nanoTime();
+      }
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:337]: Maybe `currentTimeMillis` is enough.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 6]
```diff
@@ -280,6 +280,13 @@ public void setReady(boolean ready) {
 
   private int currentWalPoolSize = 0;
 
+  // this field is used to avoid when one writer release bytebuffer back to pool,
+  // and the next writer has already arrived, but the check thread get the lock first, it find the pool
+  // is not empty, so it free the memory. When the next writer get the lock, it will apply the memory again.
+  // So our free memory strategy is only when the expected size less than the current pool size
+  // and the pool is not empty and the time interval since the pool is not empty is larger than DEFAULT_POOL_TRIM_INTERVAL_MILLIS
+  private long timeWhenPoolNotEmpty = Long.MAX_VALUE;
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:288]: Please check for grammar errors.

**Review by jt2594838 — APPROVED** [review]

**Review by jixuan1989 — APPROVED** [review]

**Pull request merged**